### DISCUSSION TITLE
Adding ENV IV (BMP280 + SHT40) to M5Stack S3 Atom Lite

### DISCUSSION THREAD

**Discussion by dxmnkd316** [ESPresense/ESPresense#1322] [Support]
I'm struggling to get the M5Stack Atom S3 Lite working with an environmental sensor via the grove port.  I can see the BMP280 in the Companion in Home Assistant, but it has a state of "Unknown".

I tried the default settings but that didn't work.  I've also tried a bunch of other bus and scl/sda settings but couldn't stumble my way into a fix either. 

I'm not sure if that's because the Atom S3 Lite has an ID of "bus_a" as shown here: https://github.com/fonske/AtomS3_lite/blob/main/ENVIV.yaml

I tried checking the "Debug I2C addreses. Look at the serial log to get the correct address" while hooked up to Arduino IDE and the ESPresense console on the firmware flashing site but it never outputs any serial info as far as I can tell.

Anyone have any idea what I'm doing wrong here?  Am I also supposed to configure something within the YAML config?

Edit:  When I'm connected to the serial monitor, I often get the following error:
"E NimBLEClient: Connection failed; status=13"

**Comment by tjharvey71** [comment]
I'm struggling with a S3 atom lite also.  The only firmware from the web page that seems to work is cdc.
I cannot figure out which pins to use for I2C based on m5stacks docs.  The esp32s3 docs say GPIO 0 and 1 but those are not available on the s3 atom lite.  I would like to connect a bme280 to mine.

I get this in my log when I select debug i2c:
 468][E][esp32-hal-i2c.c:88] i2cInit(): i2c_param_config failed

**Reply by dxmnkd316** [reply]
One of the m5 firmwares worked fine for me to get the base firmware installed. I can't remember the exact one off the top of my head.

But yeah, I think it's related to the bus ID if I had to guess.

**Comment by dxmnkd316** [comment]
Ok, I can confirm that both the S3 and the ENV IV are still functional outside of ESPresense.  I erased and flashed the S3 with just a basic example directly from M5 here:  https://github.com/m5stack/M5Unit-ENV/blob/master/examples/ENV_IV/ENV_IV.ino

It prints both the SHT and the BMP280 to the serial console just fine.  

It also looks like the /rooms/<room-name>/bmp280_temperature (and /bmp280_temperature, /humidity, and /temperature) topics are not created.  They show up in HA, but the actual MQTT topics don't show up in MQTT Explorer.  

So I have two thoughts:
1. The firmware included with ESPresense has a bug, OR
2. The ENV IV requires it's own firmware that handles both sensors

If it's 2, I suspect it comes down to including Adafruit_BMP280 in [BMP280.cpp](https://github.com/ESPresense/ESPresense/blob/master/src/BMP280.cpp) and I'm guessing isn't setup to handle the M5 ENV IV.  I could easily be wrong though since I'm still learning.

**Comment by dxmnkd316** [comment]
I was able to get the SHT40 on my ENV IV working.  I rebuilt the firmware using the esp32s3-cdc environment and changed the following lines in SHT.cpp:
14     int bus = **_1_**
33     bus = AsyncWiFiSettings.integer("SHT_I2c_Bus", 1, 2, **_1_**, "I2C Bus (-1 to disable)");
50    changed the 1 to true
51    changed the 1 to true

So forcing the bus to 1 in the firmware and possibly (though I sincerely doubt) changing the publish from 1 to true.  Though again, I doubt this had any effect.  It was just being consistent with a few of the other sensor packages.  

I used SDA = 2, SCL=1 for Bus 1 in the web config.

Still no luck on getting the pressure and temperature from the BMP280.  I tried the same changes that I did in the SHT.cpp but that didn't work.

**Comment by dxmnkd316** [comment]
@tjharvey71  - I figured it out.  Got my ENV IV reading and reporting to MQTT.
This is the first time I've been able to add something like this, so you'll have to bear with me while I get everything tested and uploaded.  I'll submit a PR in the next week or so that you can use to update your own firmware on your S3 Lite in the meantime while we wait for the PR to go through.

**Reply by tjharvey71** [reply]
Wow thats some great work!!
